### ISSUE TITLE
MVC5 Issue with Async

### ISSUE THREAD

**Issue by xxcrystalcat** [andrewdavey/NotFoundMvc#20]
I wish I knew more, I can't seem to figure this one out:

Server Error in '/' Application.
The asynchronous action method 'Register' returns a Task, which cannot be executed synchronously.
Description: An unhandled exception occurred during the execution of the current web request. Please review the stack trace for more information about the error and where it originated in the code.

Exception Details: System.InvalidOperationException: The asynchronous action method 'Register' returns a Task, which cannot be executed synchronously.

Source Error:

An unhandled exception was generated during the execution of the current web request. Information regarding the origin and location of the exception can be identified using the exception stack trace below.

Stack Trace:

[InvalidOperationException: The asynchronous action method 'Register' returns a Task, which cannot be executed synchronously.]
   System.Web.Mvc.Async.TaskAsyncActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters) +119
   System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters) +27
   System.Web.Mvc.<>c__DisplayClass12.<InvokeActionMethodWithFilters>b__f() +56
   System.Web.Mvc.ControllerActionInvoker.InvokeActionMethodFilter(IActionFilter filter, ActionExecutingContext preContext, Func`1 continuation) +256
   System.Web.Mvc.<>c__DisplayClass14.<InvokeActionMethodWithFilters>b__11() +22
   System.Web.Mvc.ControllerActionInvoker.InvokeActionMethodWithFilters(ControllerContext controllerContext, IList`1 filters, ActionDescriptor actionDescriptor, IDictionary`2 parameters) +190
   System.Web.Mvc.ControllerActionInvoker.InvokeAction(ControllerContext controllerContext, String actionName) +501
   NotFoundMvc.ActionInvokerWrapper.InvokeActionWith404Catch(ControllerContext controllerContext, String actionName) in c:\projects\opensource\NotFoundMvc\src\NotFoundMvc\ActionInvokerWrapper.cs:49
   NotFoundMvc.ActionInvokerWrapper.InvokeAction(ControllerContext controllerContext, String actionName) in c:\projects\opensource\NotFoundMvc\src\NotFoundMvc\ActionInvokerWrapper.cs:21
   System.Web.Mvc.<>c__DisplayClass22.<BeginExecuteCore>b__1e() +23
   System.Web.Mvc.Async.AsyncResultWrapper.<.cctor>b__0(IAsyncResult asyncResult, Action action) +20
   System.Web.Mvc.Async.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult) +20
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +54
   System.Web.Mvc.Controller.EndExecuteCore(IAsyncResult asyncResult) +39
   System.Web.Mvc.Controller.<BeginExecute>b__15(IAsyncResult asyncResult, Controller controller) +12
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +28
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +54
   System.Web.Mvc.Controller.EndExecute(IAsyncResult asyncResult) +29
   System.Web.Mvc.Controller.System.Web.Mvc.Async.IAsyncController.EndExecute(IAsyncResult asyncResult) +10
   System.Web.Mvc.MvcHandler.<BeginProcessRequest>b__4(IAsyncResult asyncResult, ProcessRequestState innerState) +21
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +36
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +54
   System.Web.Mvc.MvcHandler.EndProcessRequest(IAsyncResult asyncResult) +31
   System.Web.Mvc.MvcHandler.System.Web.IHttpAsyncHandler.EndProcessRequest(IAsyncResult result) +9
   System.Web.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +9514812
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +155

**Comment by frankvaneykelen** [comment]
I had the same issue with MVC 5.2.2.0. Adding the latest commit as a Project fixed it, so please do push it to NuGet!

**Comment by DavidDeSloovere** [comment]
Hi guys, the original author of NotFoundMvc have me contributor access to this repo and also the nuget package. At first I thought I could fix some things, but I'm not smart enough. So pull requests are the way to go here. I'll be glad to merge and push nuget packages (I've set up the push with myget).

**Comment by coolboy88** [comment]
why don't you fix it's nuget package?
it really takes 5 minutes!
only a small change required https://github.com/pvasek/NotFoundMvc/commit/7c2f20da25946712905a32df0c0005123f1b5e98

**Comment by DavidDeSloovere** [comment]
Please read the comments in the PR. This does not take just 5 minutes for me.  
If you wanna do a new PR, you're welcome. Make sure to add some test pages.

**Comment by mortan** [comment]
See #44 for an updated version read to be merged.

**Comment by DavidDeSloovere** [comment]
Pushed a pre-release version of 1.5 to nuget
https://www.nuget.org/packages/NotFoundMvc/1.5.0-CI00002
Who wants to test this out?

**Comment by dbruck-citytwist** [comment]
I can confirm improved behavior with NotFoundMvc version 1.5.0-CI00002 (compared to version 1.4.0).

I was using a method pair with a 'void [name]Async' and 'ActionResult [name]Completed' in an AsyncController. The [name]Async method does not call 'AsyncManager.OutstandingOperations.Increment()'; therefore, '[name]Completed' would be expected to be called synchronously afterwards.

It worked properly before installing NotFoundMvc at all, but it failed with the following error upon installing version 1.4.0:
System.InvalidOperationException: The asynchronous action method '[name]' cannot be executed synchronously.

I updated the Nuget package to pre-release version 1.5.0-CI00002 and the MVC action once again fires without error. Also, I confirmed that other basic functionality of the pre-release version appears to work as intended (rendering NotFound.cshtml when an unexpected url is entered).
